Update `Modules` and `JavaScript Compatibility Mode` docs (#1489)
grafana/k6-docs: docs/sources/next/using-k6/modules.md

import http from 'k6/http';
```

### Local modules

These modules are stored on the local filesystem, and accessed either through relative or absolute filesystem paths.

> k6 adopts a **brower-like module resolution** and doesn't support [Node.js module resolution](https://nodejs.org/api/modules.html#modules_all_together). File names for `imports` must be fully specified, such as `./helpers.js`.

Reviewer: ```suggestion
k6 adopts a **browser-like module resolution** and doesn't support [Node.js module resolution](https://nodejs.org/api/modules.html#modules_all_together). File names for `imports` must be fully specified, such as `./helpers.js`.
```

Author: thanks @yorugac

> k6 adopts a **brower-like module resolution** and does not support [Node.js module resolution](https://nodejs.org/api/modules.html#modules_all_together). File names for `imports` must be fully specified, such as `./helpers.js`.

Reviewer 2: ```suggestion
k6 adopts a **brower-like module resolution** and doesn't support [Node.js module resolution](https://nodejs.org/api/modules.html#modules_all_together). File names for `imports` must be fully specified, such as `./helpers.js`.
```

### Extension modules

Like the [k6 APIs](https://grafana.com/docs/k6/<K6_VERSION>/javascript-api), you can build custom modules in Go code and expose them as JavaScript modules. These custom Go-to-JS modules are known as [k6 extensions](https://grafana.com/docs/k6/<K6_VERSION>/extensions).

Below is an example that imports the `k6/x/kubernetes` module from the [xk6-kubernetes](https://github.com/grafana/xk6-kubernetes) extension.

```javascript
import { Kubernetes } from 'k6/x/kubernetes';

const podSpec = {
  apiVersion: 'v1',
  kind: 'Pod',
  metadata: { name: 'busybox', namespace: 'testns' },
  spec: {
    containers: [
      {
        name: 'busybox',
        image: 'busybox',
        command: ['sh', '-c', 'sleep 30'],
      },
    ],
  },
};
export default function () {
  const kubernetes = new Kubernetes();
  kubernetes.create(podSpec);
  const pods = kubernetes.list('Pod', 'testns');
  pods.map(function (pod) {
    console.log(pod.metadata.name);
  });
}
```

> How do k6 extensions (Go-to-JS modules) work? For enhanced performance, the k6 engine is written Go and embeds a JavaScript VM ([goja](https://github.com/dop251/goja)) to execute JavaScript test code. Thus, you can build your modules in Go code and import them as JavaScript as usual.

Reviewer 2: ```suggestion
How do k6 extensions (Go-to-JS modules) work? For enhanced performance, the k6 engine is written in Go and embeds a JavaScript VM ([goja](https://github.com/dop251/goja)) to execute JavaScript test code. That allows you to build your modules in Go code and import them as JavaScript as usual.
```

To learn more about using or creating k6 extensions, refer to the [Extension documentation](https://grafana.com/docs/k6/<K6_VERSION>/extensions).

## Sharing JavaScript modules

> As mentioned previously, users can import custom Javascript libraries by loading either local or remote modules. Thus, we have two options to import JavaScript modules, along with various methods to distributed them.

Reviewer 2: ```suggestion
As mentioned previously, users can import custom JavaScript libraries by loading either local or remote modules. Because of that, we have two options to import JavaScript modules, along with various methods to distribute them.
```

> > The following options for distributing and sharing JavaScript libraries are available for both custom and other public libraries.

Reviewer 2: ```suggestion
{{< admonition type="note" >}}

The following options for distributing and sharing JavaScript libraries are available for both custom and other public libraries.

{{< /admonition >}}
```

**As remote modules**

You can host your modules in a public webserver like GitHub and any CDN and be imported remotely.

```javascript
// As GitHub release assets
import {
  WorkloadConfig,
  sayHello,
} from 'https://github.com/grafana/k6-rollup-example/releases/download/v0.0.2/index.js';

// or hosted in a CDN
import { randomIntBetween, randomItem } from 'https://jslib.k6.io/k6-utils/1.4.0/index.js';
```

When the library consists of multiple files and modules, you may want to bundle these modules to create public releases. Here are some examples for reference:

- Using Webpack: [k6-jslib-utils](https://github.com/grafana/k6-jslib-utils) and [k6-jslib-aws](https://github.com/grafana/k6-jslib-aws).
- Using Rollup: [test-commons](https://github.com/grafana/k6-rollup-example/).

Be aware that k6 automatically executes remote modules, making it crucial to trust the source code of these remote modules. There is a **risk of altering the remote modules with certain hosting mechanisms**. To mitigate this security risk, some users prefer to download and import the modules locally to ensure full control of the source code.

**As local modules**

> In this example, we've downloaded the previous remote modules in the `lib` folder of the testing project and imported them as follows:

Reviewer 2: ```suggestion
In this example, the previous remote modules have been downloaded to the `lib` folder of the testing project and imported as follows:
```

{{< /code >}}

Note that on Windows, you also need to make sure that your drive in question, say `C:\`,
has been marked for sharing in the Docker Desktop settings.

### Extension modules

The official [Grafana k6 Docker image](https://hub.docker.com/r/grafana/k6) includes the k6 release binary but lacks additional k6 extensions. Therefore, using the official Docker container to run a k6 test that requires an extension will fail.

> To run k6 with any extension in Docker, you must create a custom Docker image that includes the k6 binary with the necessary extensions. To learn building a Docker image with k6 and extensions, refer to the [xk6 command](https://grafana.com/docs/k6/<K6_VERSION>/extensions/build-k6-binary-using-go#breaking-down-the-xk6-command) and this [Dockerfile](https://github.com/grafana/xk6-output-prometheus-remote/blob/main/Dockerfile) example.

Reviewer 2: ```suggestion
To run k6 with any extension in Docker, you must create a custom Docker image that includes the k6 binary with the necessary extensions. To learn how to build a Docker image with k6 and extensions, refer to the [xk6 command](https://grafana.com/docs/k6/<K6_VERSION>/extensions/build-k6-binary-using-go#breaking-down-the-xk6-command) and this [Dockerfile](https://github.com/grafana/xk6-output-prometheus-remote/blob/main/Dockerfile) example.
```

{{< /code >}}

Note that on Windows, you also need to make sure that your drive in question, say `C:\`,
has been marked for sharing in the Docker Desktop settings.

### Extension modules

The official [Grafana k6 Docker image](https://hub.docker.com/r/grafana/k6) includes the k6 release binary but lacks additional k6 extensions. Therefore, using the official Docker container to run a k6 test that requires an extension will fail.

To run k6 with any extension in Docker, you must create a custom Docker image that includes the k6 binary with the necessary extensions. To learn building a Docker image with k6 and extensions, refer to the [xk6 command](https://grafana.com/docs/k6/<K6_VERSION>/extensions/build-k6-binary-using-go#breaking-down-the-xk6-command) and this [Dockerfile](https://github.com/grafana/xk6-output-prometheus-remote/blob/main/Dockerfile) example.

```bash
# Stage 1 - download xk6 and desired extensions, then compile a new binary.

> FROM golang:1.20-alpine3.18 as builder

Reviewer 3: Not certain we want to have this in docs and then forget to update them all the time.

Maybe linking to xk6 repo and it docs will be better?

We do bump go version all the time. Go 1.20 is no logner supported from today as 1.22 got released and we already mostly build with 1.21. 

Similarly alpine is on version 3.19

Reviewer 4: The minimal Dockerfile to produce an image with custom `k6` build would be

```Dockerfile
FROM grafana/xk6:latest

RUN GCO_ENABLED=0 xk6 build \
    --with github.com/grafana/xk6-output-prometheus-remote@v0.3.1

ENTRYPOINT ["./k6"]
```

If necessary, one could implement a multistage build and copy the `xk6 build` artefact, but I'd leave it out to users.

Reviewer: A related thought: is it possible to source the files from repo to the doc? I.e. to have a full file inserted when docs are published. IMHO, in this case, it is better to have a full listing instead of a link to repo and it would also result in the latest instruction set, without the need to constantly update it.
I suppose cc @heitortsergent ?

Author: https://github.com/grafana/k6-docs/pull/1489/commits/95fb2e43a1cb2d81d37285d54a80c093594fd320 adds @andrewslotin instructions.

These instructions are very easy to follow ❤️. They show quickly the `xk6 build` command, which is the core topic we want users to learn for their particular case. 

>  is it possible to source the files from repo to the doc? 

@heitortsergent @yorugac, please, let's discuss this separately. 

I am resolving this conversion. Please, reopen if it needs further discussion.

Reviewer 2: I opened https://github.com/grafana/k6-docs/issues/1508 so we can discuss this separately. :)

I know we have the ability to import files from different repos, but I'm not sure about the details.
